bindings/gumjs/gumdukcore.c

```diff
@@ -2816,50 +2814,85 @@ gum_duk_native_function_invoke (GumDukNativeFunction * self,
                                 duk_idx_t argv_index)
 {
   GumDukCore * core;
-  gsize nargs;
+  ffi_cif *cif;
   ffi_type * rtype;
+  ffi_type ** atypes;
+  gsize nargs, nargs_fixed;
+  gboolean is_variadic;
   gsize rsize, ralign;
   GumFFIValue * rvalue;
   void ** avalue;
   guint8 * avalues;
+  ffi_cif tmp_cif;
+  GumFFIValue tmp_value = {0};
   GumDukSchedulingBehavior scheduling;
   GumDukExceptionsBehavior exceptions;
   GumDukReturnValueShape return_shape;
   GumExceptorScope exceptor_scope;
   gint system_error;
 
   core = self->core;
-  nargs = self->cif.nargs;
-  rtype = self->cif.rtype;
+  cif = &self->cif;
+  nargs = cif->nargs;
+  nargs_fixed = self->nargs_fixed;
+  is_variadic = self->is_variadic;
 
-  if (argc != nargs)
+  if ((is_variadic && argc < nargs_fixed) || (!is_variadic && argc != nargs))
     _gum_duk_throw (ctx, "bad argument count");
 
+  rtype = cif->rtype;
+  atypes = cif->arg_types;
   rsize = MAX (rtype->size, sizeof (gsize));
   ralign = MAX (rtype->alignment, sizeof (gsize));
   rvalue = g_alloca (rsize + ralign - 1);
   rvalue = GUM_ALIGN_POINTER (GumFFIValue *, rvalue, ralign);
 
-  if (nargs > 0)
+  if (argc > 0)
   {
-    gsize arglist_alignment, offset, i;
+    gsize arglist_size, arglist_alignment, offset, i;
 
-    avalue = g_alloca (nargs * sizeof (void *));
+    avalue = g_alloca (MAX (nargs, argc) * sizeof (void *));
 
-    arglist_alignment = self->cif.arg_types[0]->alignment;
-    avalues = g_alloca (self->arglist_size + arglist_alignment - 1);
+    arglist_size = self->arglist_size;
+    if (is_variadic && argc > nargs)
+    {
+      gsize type_idx;
+
+      atypes = g_newa (ffi_type *, argc);
+
+      memcpy(atypes, cif->arg_types, nargs * sizeof (ffi_type *));
+      for (i = nargs, type_idx = nargs_fixed; i != argc; i++)
+      {
+        ffi_type * t = cif->arg_types[type_idx];
+
+        atypes[i] = t;
+        arglist_size = GUM_ALIGN_SIZE (arglist_size, t->alignment);
+        arglist_size += t->size;
+
+        if (++type_idx >= nargs)
+          type_idx = nargs_fixed;
+      }
+
+      cif = &tmp_cif;
+      if (ffi_prep_cif (cif, self->abi, (guint) argc, rtype,
```

Comment:
Is this correct? Some ABIs, e.g. iOS/arm64, treat non-variadic and variadic functions very differently, so if we don't tell libffi that the function is variadic, it's not going to be able to call it correctly.

Reply:
This is copied from pyobjc.  I searched the source of pyobjc, it do not use the  `ffi_prep_cif_var` and all use `ffi_prep_cif`. Maybe pyobjc is only target on MacOSX, so that is OK, but may be wrong for  iOS/arm64. I will correct it.